### DISCUSSION TITLE
API function return undefined request body (Content-Type: Image/png)

### DISCUSSION THREAD

**Discussion by NeikiDev** [vercel/community#1216] [Help] [answered]
### Summary

```js
const handler = async (request, response) => {
        console.log(request.headers) // normal headers
        console.log(request.body) // undefined
       }
```
![Code_UxwAZ062gA](https://user-images.githubusercontent.com/67499407/211048265-a191ae2b-9694-4522-b352-78325f5352cd.png)

### Additional information

Hey, i switch some weeks ago to vercel and rewrote my API functions, so i have an uploader api that takes as request.body the image data as binary and push it to github threw oktokit so the website gets redeployed with the new image, all is working, the push and stuff, but if i use f.ex and image uploader like ShareX the "request.body" is "undefined", i dont know why this happens (tested locally with dev mode and normal my publish api) both times its "undefined" if the "content-type" is set to "image/png", other times if i send raw data with postman to test and the content type is text then the raw input is there.

(The upload to Github takes base64 as content so i need that raw image data)

**Comment by amyegan** [comment]
My first thought was that you could be running into a [Serverless Function Payload Size Limit](https://vercel.com/docs/concepts/limits/overview#serverless-function-payload-size-limit), but then you would be getting a `413: FUNCTION_PAYLOAD_TOO_LARGE` error. Do you have a public repo or [minimal reproduction](https://vercel.com/support/articles/creating-a-minimal-reproducible-example) that I can use to gather more information about the problem?

**Reply by NeikiDev** [reply]
Hey, you can test it using this endpoint here on my website (currently only for testing) "https://neiki.dev/api/upload" 

It returns the request body,method,content-type,content-length

Im using the program ShareX to send an Image data (binary) with an post request to that endpoint above, if the ContentType is an Image/png the body is undefined , if i set it to text/plain, i get the raw data

im really confused why this is happening 😅

**Reply by NeikiDev** [reply]
Here you can see 2 Examples i am testing, im confused tbh

1: ContentType to Image
2: ContentType to Text

the other stuff is the same

![ShareX_xuKyCvn2lE](https://user-images.githubusercontent.com/67499407/211058345-656c19c4-5b6f-4e21-bdba-9b9eec5a30a4.png)
![ShareX_8MJlquHhTu](https://user-images.githubusercontent.com/67499407/211058584-053cae46-46f9-4bf8-a9ef-d97cf808f03a.png)

**Reply by NeikiDev** [reply]
My First website was hosted on another host with that type of functions, and i used it there and got the image data as base64 exactly the same directly without any problems 🤔 thats why i am so confused, that the body is just "undefined"

Some weird thing too:

```js
console.log(request) // body: Getter/Setter
console.log(request.body) // undefined
console.log(request) // body: undefined
```

if i do this in that row just get that

**Comment by NeikiDev** [comment, marked as answer]
Okay i now fixed it ig on a maybe weird way

Solution:

```js
async function buffer(readable) {
    const chunks = [];
    for await (const chunk of readable) {
        chunks.push(typeof chunk === 'string' ? Buffer.from(chunk) : chunk);
    }
    return Buffer.concat(chunks);
}

const handler = async (request, response) => {
        const buf = await buffer(request);
        const bas64Body = buf.toString('base64');
        
        uploadBase64ToGithub(base64Body)
        
        ....
        
```

im not sure how it actually works, but it worked. The image is now uploaded and now errors nothing, valid Base64 Data and everything 🤔 the request.body still undefined tho.
If somebody got another solution or something, or maybe can explain it to me. please reply here 😊
